File: JS/sorting/selection_sort.js
/******************************* SELECTION SORT ******************************/
/* principle -> 
    first make the first element the smallest element and compare with all other elements, if smaller number is found then make it smallest number. continue this process and push smallest element of every iteration in new array.
*/
const selectionSort = (arr) => {
    let smallest = arr[0]
    let smallestIndex = 0;
    for (let i = 1; i < arr.length; i++) {
        if (arr[i] < smallest) {
            smallest = arr[i]
            smallestIndex = i;
        }
    }
    return smallestIndex
}

const sorting = (arr) => {
    const newArr = []
    const length = arr.length
    for (let i = 0; i < length; i++) {
        let smallestIndex = selectionSort(arr)
        newArr.push(...arr.splice(smallestIndex, 1))
    }
    console.log(newArr)
}
sorting([5, 3, 6, 2, 10]) //BigO -> O(n*n) it actually is  O(n x n x 1/2) but we ignore the constants bestCase -> O(n*n)

/******************************* SELECTION SORT ******************************/

File: selectionsort.py
#program for selection sort using python 
def selectionSort(array, size):
    
    for ind in range(size):
        min_index = ind
 
        for j in range(ind + 1, size):
            # select the minimum element in every iteration
            if array[j] < array[min_index]:
                min_index = j
         # swapping the elements to sort the array
        (array[ind], array[min_index]) = (array[min_index], array[ind])
 
arr = [46,57,38,79,13,76,32,120]
size = len(arr)
selectionSort(arr, size)
print('The array after sorting in Ascending Order by selection sort is:')
print(arr)
